### ISSUE TITLE
Branching scheme for `apigentools push`

### ISSUE THREAD

**Issue by masci** [DataDog/apigentools#43]
**Overview of issue:**

The repos hosting generated clients have no `master` branch after the first `apigentools push` and the proposed PRs  are created against a weirdly named branch instead of `master`, see for example (sorry the repo isn't publicly available yet, can't link PR directly):

![image](https://user-images.githubusercontent.com/7241/64550034-4de21b00-d332-11e9-98c4-2bd6a77e6a4c.png)

I didn't notice this behaviour until a colleague wanted to fix a doc issue on a client and couldn't find `master` when opening a PR, so I guess this is an issue only if you intend to push code to a repo with both `apigentools push` and manual work from contributors.

**Steps to reproduce the issue:**
1. create an empty repo for a client, `master` is there
1. run `apigentools generate --builtin-templates --clone-repo`
1. run `apigentools push`
1. `master` is gone, a new branch named `lang/xxxx` is the only branch present
1. change the spec
1. repeat gen/push steps
1. since `master` is gone, PR to update the client wants to merge into `lang/xxxx`

**Comment by bkabrda** [comment]
Hey, thanks for the report. This might be an issue with how Github creates and treats repos with empty master branches. I'll do some experiments tomorrow and try to send a PR to fix this.

As a temporary workaround, I'd suggest manually pushing a master branch and setting it as a default branch [1] - that should make sure you won't hit this issue with that particular repo in the future.

[1] https://help.github.com/en/articles/setting-the-default-branch

**Comment by bkabrda** [comment]
I just did some experiments and here's more detailed notes:

This would get triggered when an empty repo is created on Github; empty Git repos don't have *any* branches (`git branch` doesn't print anything when you clone). So when `apigentools push` is invoked first time, it will push the automatically created branch `{language}/{timestamp}`. Github will then set this branch as default, since it's the first branch created in the repository. A subsequent push will create a new branch and when a PR is opened, it will be opened against the now-default branch.

So in your case the `master` branch is not deleted AFAICS, it was never created in the first place.

The question is how to solve this. I think the best way to go here is for the `push` command to fail when `master` branch doesn't exist in the remote repository (plus providing a good error message and an argument to optionally disable this new behavior). WDYT @masci? (also CC @nmuesch @gzussa @zippolyte for more opinions on this)

**Comment by zippolyte** [comment]
At this point, shouldn't we simply create the `master` branch when we detect there are no branches in the repo instead of failing ? So that everything is transparent to the user. 

When there are no branches, I don't think it's a problem to assume we can just create `master` and make it default, it means the repo is completely new.

**Comment by masci** [comment]
@bkabrda you're right, I've messed up the reproduce steps, the `master` isn't there right after creating an empty repo, this explains why I couldn't find the point where apigentools was "wiping" `master` :)

Technically speaking, I think the tool is behaving correctly but it seems you pointed out two interesting solutions that would improve user experience and that I'd recap like this:

1. Fail push when there are no branches in the repo with an error suggesting to create one
1. Create and push a default branch when there are no branches in the repo, name could be `master` or whatever makes more sense

Both would have prevented the issue, it's more a matter of style at this point, depending on whether you're more comfortable to better ask permission than forgiveness (n.1) or the other way around (n.2).

**Comment by nmuesch** [comment]
Thanks for investigating this. I'm not sure how common this scenario is, but wanted to note it. Both solutions assume that the user wants to name their default branch `master`. We could make that name configurable, perhaps in the config of the spec repo like we have for the github org and repo names.

I'm personally more in favor with the first proposed solution (of raising an error if the configured default branch is missing), mostly because I don't think there is a way to set a branch as default through `git` since that's a Github thing.

**Comment by zippolyte** [comment]
I don't think making it configurable would be useful with solution 1. The point is you should raise an error only if your repo is empty and does not have any branch. Once that's done, then you're all set, `apigentools push` will just have its regular behavior. In other words, no need to check that a specific branch is present, you just check whether or not there is a branch at all.

With solution 2, yeah why not making that default branch name configurable with `master` being the default name. That'd be a shortcut for manually creating and pushing a first branch to your newly created repository.

Now, the default branch can be changed in GitHub after the fact easily, so I'd rather do solution 2, just to have a better user experience when using `apigentools`. Getting an error on the first try doesn't sound great. 

> I don't think there is a way to set a branch as default through git since that's a Github thing.

Yes there is: you push it first

**Comment by nmuesch** [comment]
> I don't think there is a way to set a branch as default through git since that's a Github thing.
I should have clarified that I meant once any branch is pushed to the repository you can't change the default via `git`. 

Yeah in that case I'm good with option 2 if we have the default branch configurable in the apigentools config.

**Comment by bkabrda** [comment]
Thanks for your opinion folks. I think from the discussion, we will:

* Add an argument to specify the "main" branch (defaulting to master)
* We'll assume that if the "main" branch is not present in the repo, someone is initializing the repo and so we'll do the commit on the main branch instead of the feature branch.

Does that sound good?
